### PR TITLE
feat(composer): link editing — inline popover, modal picker, buzz:// in-app nav

### PR DESCRIPTION
**Pull request by tellaho**
**Category:** improvement

**User Impact:** The composer's link button now works (it was silently broken in the desktop app), and you can click an existing link to see, edit, or remove it via an inline popover — without being pulled out of the editor. `buzz://` deep-links open in-app, just like rendered messages.

**Problem:** Highlighting text and clicking the link icon did nothing — the handler called the native URL prompt, a no-op in the Tauri WebView. There was also no way to see, edit, or remove the URL behind a set link (the link extension runs with `openOnClick: false` and had no affordance). A first pass surfaced editing through a takeover modal, which was jarring when a user only wanted to tweak display text inline.

**Solution:** Replace the native prompt with a shadcn dialog (display text + URL) for adding links, and add an inline, information-only popover for existing links — display text, the URL as a clickable hyperlink, and Edit/Remove actions — that keeps the user in the editor. Link interactions (click interception, click-vs-drag selection preservation, and active-link reporting as the cursor moves) are centralized in a ProseMirror plugin. The popover's URL routes `buzz://message?…` deep-links in-app via the same path the rendered-message renderer uses; everything else opens in the OS browser. Add-link focuses the URL field; editing an existing link focuses display text.

<details>
<summary>File changes</summary>

**desktop/src/features/messages/lib/linkInteractionExtension.ts** (new)
ProseMirror plugin centralizing composer link interactions: click interception, click-vs-drag selection preservation, and active-link reporting as the selection moves (drives the cursor-anchored link card).

**desktop/src/features/messages/lib/useLinkEditor.tsx**
Owns the link-edit modal and the inline link-info popover — seeds from toolbar selection, a clicked link, or the active cursor link; handles add/edit/remove. Routes the popover URL through the canonical link-open path.

**desktop/src/features/messages/lib/resolveLinkAt.ts**
Resolves the full ProseMirror link mark range under a position by walking contiguous siblings sharing the same href (covers links spanning mixed formatting).

**desktop/src/features/messages/lib/linkEditorFocus.ts** (new)
Pure helper deciding initial focus: selected-text link insert → URL field; otherwise → display-text field.

**desktop/src/features/messages/lib/useRichTextEditor.ts**
Wires the link-interaction extension and the link-selection/apply/remove helpers into the editor.

**desktop/src/features/messages/ui/MessageComposer.tsx** / **desktop/src/features/forum/ui/ForumComposer.tsx**
Wire the link editor + popover into both composers consistently.

**desktop/src/shared/ui/button.tsx**, **desktop/src/shared/styles/globals.css**
Minor supporting tweaks for the popover action layout.

</details>

## Reproduction Steps
1. Run the desktop app, open a channel.
2. Type text, highlight it, click the link icon → the add-link modal opens with the URL field focused (previously: nothing happened).
3. Enter a URL, Save → the highlighted text becomes a link.
4. Click the set link → an inline popover shows the display text + URL (as a clickable link) with Edit and Remove; editor focus is preserved. Edit opens the modal focused on display text; Remove strips the hyperlink.
5. Click a `buzz://message?…` link's URL in the popover → navigates in-app to that channel/message; an `https://` URL opens in the OS browser.
6. Set a link whose text carries mixed formatting (bold/italic) → Edit/Remove resolves the full range, no dangling half-link.

Screen Recording 2026-06-14 at 3 36 53 PM

|  |  |
|---|---|
|popover|image|
|edit modal|image|
|add modal|image|

### CODE REVIEW

**Review by wesbillman — COMMENTED** [review]
Thorough review from Brain + Pinky (split: core link logic / UI-lifecycle). Overall this is solid, well-decomposed work — pure helpers extracted and unit-tested, ProseMirror concerns isolated in the plugin, and the popover URL routing genuinely reuses the rendered-message path instead of forking. **Ship-worthy after the small fixes below.** Inline comments mark the one real bug (a node-boundary edge case in `resolveLinkAt`) plus a couple of readability nits.

File: desktop/src/features/messages/lib/resolveLinkAt.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  type ChildSpan = { from: number; to: number; hasLink: boolean };
+  const spans: ChildSpan[] = [];
+  let anchorIndex = -1;
+  parent.forEach((child, childOffset) => {
+    const childFrom = parentStart + childOffset;
+    const childTo = childFrom + child.nodeSize;
+    const childLink = linkType.isInSet(child.marks);
+    const hasLink = childLink != null && childLink.attrs.href === href;
+    if (childFrom <= pos && pos <= childTo) anchorIndex = spans.length;
```
Comment (wesbillman) [inline desktop/src/features/messages/lib/resolveLinkAt.ts]: **Node-boundary double-match (real, edge case).** Both ends of this test are inclusive (`childFrom <= pos && pos <= childTo`), and span N's `childTo` equals span N+1's `childFrom`. When `pos` lands exactly on the seam between two nodes, *both* spans match and the second wins (last assignment). For two adjacent links with different hrefs, a click/caret on the seam anchors on the wrong link.

Fix: make the right edge exclusive for the anchor pick (`pos < childTo`), or prefer the span whose `hasLink && href === mark.href`. The existing "does not extend across an adjacent link with a different href" test clicks mid-node (pos 3), so it doesn't exercise the seam — worth adding a boundary case (and a "caret at the very start of a link" case to lock the `pos + 1` left-edge fallback above).

File: desktop/src/features/messages/lib/resolveLinkAt.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  // The mark at the caret sits on the character *before* the position, with
+  // the character after as a fallback (caret at a link's left edge).
+  const mark =
+    linkType.isInSet($pos.marks()) ||
+    (pos < state.doc.content.size
+      ? linkType.isInSet(state.doc.resolve(pos + 1).marks())
+      : null);
+  if (!mark) return null;
+
+  const href = mark.attrs.href as string;
```
Comment (wesbillman) [inline desktop/src/features/messages/lib/resolveLinkAt.ts]: Nit: `mark.attrs.href as string` — if the schema guarantees `href` is a string this is fine; otherwise a runtime guard wouldn't hurt. Non-blocking.

File: desktop/src/features/messages/lib/useLinkEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+
+  const close = React.useCallback(() => setDraft(null), []);
+
+  const closeCard = React.useCallback(() => setCardState(null), []);
+
+  const save = React.useCallback(() => {
+    if (!draft) return;
+    const url = draft.url.trim();
+    if (!url) return;
+    if (draft.from === 0 && draft.to === 0) {
```
Comment (wesbillman) [inline desktop/src/features/messages/lib/useLinkEditor.tsx]: Nit: magic sentinel. `from === 0 && to === 0` works because PM doc positions are 1-based (0 is never a valid content position), but it reads as a magic number — and it's coupled to the `from: 0, to: 0` set in `openFromToolbar`. A named `hasRange` boolean (or a dedicated `null` range) on the draft would make the intent obvious. Also note: when this fallback hits and `getLinkSelectionInfo()` returns `null` (empty caret), `from`/`to` resolve to `0` — a genuine non-content position. Worth a quick sanity check that insert-at-empty-caret lands where expected. Non-blocking.

File: desktop/src/features/messages/lib/resolveLinkAt.ts (hunk 1 above)
Comment (tellaho) [reply desktop/src/features/messages/lib/resolveLinkAt.ts]: Good catch on the seam ambiguity. Reworked the resolver to anchor on the character *before* the caret (`$pos.nodeBefore`), falling back to `nodeAfter` for the left edge and `$pos.marks()` mid-node. The span anchor is now two-pass: an exclusive right edge (`pos < childTo`) plus a preference for the span that both contains `pos` and carries the target href, so the left link wins at a seam between two different links. Added two boundary tests — seam-between-adjacent-links and caret-at-very-start. Fixed in 52bdad0f. 🤖

File: desktop/src/features/messages/lib/useLinkEditor.tsx (hunk 3 above)
Comment (tellaho) [reply desktop/src/features/messages/lib/useLinkEditor.tsx]: Replaced the magic `from === 0 && to === 0` sentinel with an explicit `hasRange` boolean on the draft state, set at all three draft-creation sites. `save` now branches on `hasRange`, and I pushed the range-fallback logic down into `applyLink` (which holds the editor state): it takes an optional range and defaults to the live PM selection, so an empty-caret insert lands at the cursor rather than placeholder doc position 0. Fixed in ad2625ee. 🤖

File: desktop/src/features/messages/lib/resolveLinkAt.ts (hunk 2 above)
Comment (tellaho) [reply desktop/src/features/messages/lib/resolveLinkAt.ts]: Replaced the `as string` cast with a runtime `typeof` guard — a malformed link mark now resolves to `null` instead of letting a non-string href flow through the resolver. Fixed in e7592e31. 🤖

**Comment by wesbillman** [comment]
@codex please review

**Review by wesbillman — APPROVED** [review]
